src/components/application_manager/rpc_plugins/vehicle_info_plugin/include/vehicle_info_plugin/custom_vehicle_data_manager_impl.h

```diff
@@ -0,0 +1,86 @@
+#ifndef SRC_COMPONENTS_APPLICATION_MANAGER_RPC_PLUGINS_SDL_RPC_PLUGIN_INCLUDE_SDL_RPC_PLUGIN_VEHICLE_DATA_VALIDATION_MANAGER_IMPL_H
+#define SRC_COMPONENTS_APPLICATION_MANAGER_RPC_PLUGINS_SDL_RPC_PLUGIN_INCLUDE_SDL_RPC_PLUGIN_VEHICLE_DATA_VALIDATION_MANAGER_IMPL_H
+
+#include <boost/optional.hpp>
+
+#include "smart_objects/smart_object.h"
+#include "vehicle_info_plugin/custom_vehicle_data_manager.h"
+
+namespace rpc {
+namespace policy_table_interface_base {
+class VehicleDataItem;
+}
+}  // namespace rpc
+
+namespace policy {
+class VehicleDataItemProvider;
+}
+
+namespace vehicle_info_plugin {
+namespace policy_table = rpc::policy_table_interface_base;
+
+typedef boost::optional<const policy_table::VehicleDataItem> OptionalDataItem;
+typedef std::set<std::string> VehicleInfoSubscriptions;
+
+class CustomVehicleDataManagerImpl : public CustomVehicleDataManager {
```

Comment:
With pimple Idiom, we will also need to declare function 3 times, 1 time in header and 2 times in cc file.
I am trying to keep consistency with whole project (using impl classes). 
Btw, I will recreate that class in pimple style and probably it will looks better.

Reply:
@jacobkeeler  I did some investigations and several solutions and can't see code improvements with using pimple idiom here. 

We need to make Pimple mockable, so the initial class should have an interface, so we are not reducing the number of places where the function is declared.  

Another solution is to mock `CustomVehicleDataManager::Impl` class but it leads to many potential linkage and runtime errors.

Anyway, I do not see the benefits of introducing Pimpl idiom in class that is platform agnostic.

Please see attempt to introduce pimple in this PR https://github.com/LuxoftAKutsan/sdl_core/pull/282

Reply:
You shouldn't need to mock the implementation class, just the interface.

One misunderstanding that I see with the attempted implementation that you linked is that you have `CustomVehicleDataManagerImpl` (which should be `CustomVehicleDataManager::Impl`) as a subclass of `CustomVehicleDataManager`, when it should just be a collection of implementation details for the class, and should not be a subclass of anything. You should not need a separate `create` function, this object should be created in the normal constructor.

custom_vehicle_data_manager.h
```
  private:
    std::shared_ptr<CustomVehicleDataManager::Impl> impl_;
```
custom_vehicle_data_manager.cc
```
  CustomVehicleDataManager::CustomVehicleDataManager() {
    impl_ = new CustomVehicleDataManager::Impl();
  }
```

It seems to me like there would be no issue with mocking the `CustomVehicleDataManager` in this case.

Reply:
But what problem do we resolve in that case?
All functions any way will be repeated multiple times :

 - In interface
 - in implementation
 - in pimple
 - in mock

Consider : 
‘’’
Class Foo {
 virtual void foo() =0;
};

Class FooImpl : Foo {
 struct Impl;
  Void foo() override;
Impl* impl_;
}

Class FooImplImpl { // smels
  Void foo() {...}
}

FooImpl::foo(){ impl_->foo();}

FooMock : Foo {
  Mockmethod0(foo, void());
}
‘’’

If we will skip FooImpl, and put Impl class into Foo, Mock will inherit behavior of  implementation class (throw constructor). And if some method will no be mocked, real implementation will be called (without warnings)
I feel it as bad practice in unit tests, Mocked classes should not contain implementation, and errors about missing mocked methods should present in compile time

Reply:
The major downside to the current approach is that each function that is added or changed requires you to modify the interface header, implementation header, and implementation file for every affected class. This can add up significantly, notably when a function is propagated upward (such as with any cache manager functions).

In addition, I find it to be an anti-pattern to create an interface which is only meant to be inherited by a single class. I have realized that there are significant benefits to separating public class details and private implementation details, and the "PImpl" idiom seems to be the only way to do so without the downsides mentioned above. Both approaches appear to have a similar amount of code, this approach just uses fewer files.

Reply:
What approach do you propose to Mock behavior of Pimpl classes without an interface?

Reply:
Impl functions are private to the implementation of the class, you absolutely should not be defining the same function in both `Foo` and the `Foo::Impl`. The example you gave is incorrect, it would be rewritten as such:
```
// foo.h
Class Foo {
 public:
  virtual void publicFunc() = 0;
 private:
  class Impl;
  std::unique_ptr<Impl> impl_;
}

// foo.cc
class Foo::Impl {
  std::string privateMember;
  void privateFunc() {...}
}

Foo::publicFunc() {
  ... function contents ...
  impl_->privateFunc();
  ... more function contents ...
}

// foo_mock.h
FooMock : Foo {
  Mockmethod0(publicFunc, void());
}
```

The Impl class basically replaces all private members and functions of the class. That way, only public details are defined in the header. You should only need to mock public functions, so there should be no issue here.

Reply:
@jacobkeeler  `virtual void publicFunc() = 0;`

 - In that case we will not be able to create instance of Foo.
 - FooMock will contain Foo::Impl instance
 - FooMock may miss mocking some methods, so real implementation may be called.

I feel uncomfortable with this issues.

Reply:
> If we will skip FooImpl, and put Impl class into Foo, Mock will inherit behavior of implementation class (throw constructor). And if some method will no be mocked, real implementation will be called (without warnings)
> I feel it as bad practice in unit tests, Mocked classes should not contain implementation, and errors about missing mocked methods should present in compile time

As long as each of the public functions are `virtual` in the header, there should be no practical issue with this, since the mock function would completely override the implementation behavior (see the `AppServiceManager` class). I do see how this could be seen as a bad practice, but this is just an unfortunate limitation of GMock (or maybe c++ as a whole) when attempting to mock concrete classes. I believe that the inclusion of unnecessary interface classes is a worse practice in the long run. 

I would add the caveat that this should only be done for concrete classes which will only ever have one implementation, as is the case for any class that currently has the `Impl` suffix. If there was a practical reason for including an interface class (meaning that it would be practical for multiple classes to inherit from it), then it should be included, but this is definitely not one of those cases.

Reply:
@LuxoftAKutsan 
* Seems that this was mistake in syntax on my part, the `= 0` would be removed, I will update the original comment. See `AppServiceManager` for an example of a mocked concrete class.
* I don't see this as a massive issue but admittedly I am surprised there isn't a reasonable workaround for this. I'd be even less concerned if the pointer remained empty, but I don't see a way to do that at the moment.
* I suppose `NiceMock` is a workaround for this, but I do now see this as a legitimate concern.

Investigating the issues, I am starting to realize that there really isn't a proper solution to this which incorporates testing in c++, unfortunately. I really wish that we could properly do away with the `Impl` class pattern, but I am willing to stop pushing for this at this point since there are a few legitimate concerns with the approach I suggested.

Reply:
I will think about the problem you specified, I agree that it is ugly to modify so many files if we should add a new public method. But I think we should do it as separate refactoring on already existing classes to check that is does not introduce any regression and does not change the behavior of the system.